### PR TITLE
Add /cohorts/{id}/users endpoint

### LINKED ISSUE #31: API: User table - GET - Get all users by cohort ID
Description
Get all users with cohort ID: Implement functionality to get all users from the database with a specific cohort ID attribute. This functionality is required for the dashboard "my cohorts" functionality in the frontend

**route: **
GET - user/cohort/{cohortId}

RESPONSE:
->status
200 OK, 401 Fail

Response body example:
```
{
  "status": "string",
  "data": {
    "users": [
      {
            "id": 0,
            "email": "string",
            "cohort_id": 0,
            "role": "string",
            "firstName": "string",
            "lastName": "string",
            "bio": "string",
            "githubUsername": "string",
            "mobile": "string",
            "specialism": "string",
            "imageUrl": "string"
          }
    ]
  }
}
```

### PR DESCRIPTION
**Pull request by gremble0**
https://github.com/orgs/boolean-uk/projects/19/views/2?pane=issue&itemId=85334908&issue=boolean-uk%7Cteam-dev-server-ex-2410-team1%7C31

### CODE REVIEW

**Review by JosteinRuen — DISMISSED** [review]
Great addition

**Review by alinjo — DISMISSED** [review]
just barely;)

File: docs/openapi.yml [hunk 1]
```diff
@@ -404,6 +404,37 @@ paths:
               schema:
                 $ref: '#/components/schemas/Error'
 
+  /cohorts/{id}:
```
Comment (dogezen) [inline docs/openapi.yml]: this `GET /cohorts/{id}` tells me that the endpoint should be GET Cohort data by ID (ie. all cohort data: users, cohort name, start date, etc...), not just the users.

if you want to just do GET cohort users, then the endpoint route should be: `GET /cohorts/{id}/users`

File: src/controllers/cohort.js [hunk 2]
```diff
@@ -10,3 +16,23 @@ export const create = async (req, res) => {
     return sendMessageResponse(res, 500, 'Unable to create cohort')
   }
 }
+
+/**
+ * Updates a user by ID
+ * @param {import('express').Request} req
+ * @param {import('express').Response} res
+ */
+export const getById = async (req, res) => {
```
Comment (dogezen) [inline src/controllers/cohort.js]: rename to `getCohortUsers` or `getCohortUsersByCohortId`

File: src/controllers/cohort.js [hunk 3]
```diff
@@ -10,3 +16,23 @@ export const create = async (req, res) => {
...
+  // TODO: not User.findAll()
+  const id = parseInt(req.params.id)
+
+  try {
+    const allUsers = await User.findAll()
```
Comment (dogezen) [inline src/controllers/cohort.js]: what if we have 1000 users, you'll go to DB, get all 1000 rows and then filter out here, better to directly filter before sending the query using the prisma `where` condition.

You could add a `Cohorts` domain file, that has a function `Cohort.getUsers(cohortId)` -> this then internally calls prisma with the correct where clause

File: src/controllers/cohort.js [hunk 4]
```diff
@@ -10,3 +16,23 @@ export const create = async (req, res) => {
...
 }
+
+/**
+ * Updates a user by ID
+ * @param {import('express').Request} req
+ * @param {import('express').Response} res
+ */
+export const getById = async (req, res) => {
+  // TODO: not User.findAll()
+  const id = parseInt(req.params.id)
```
Comment (dogezen) [inline src/controllers/cohort.js:25]: at the moment, there's no validation that the cohort with this id exists, not sure if you want to return 404 for non-existent cohorts

File: src/routes/cohort.js [hunk 5]
```diff
@@ -8,5 +8,6 @@ import {
 const router = Router()
 
 router.post('/', validateAuthentication, validateTeacherRole, create)
+router.get('/:id', validateAuthentication, getById)
```
Comment (dogezen) [inline src/routes/cohort.js]: change to `/:id/users` since we are getting the cohort's users and not the whole cohort data

**Review by dogezen — COMMENTED** [review]
Please see comments, it seems you want to fetch JUST the cohort users, instead of the whole cohort data, so I have added comments accordingly.

I'd suggest adding a Cohort domain now since the prisma call should happen on the domain side (see comments)

File: docs/openapi.yml (hunk 1 above)
Comment (gremble0) [reply docs/openapi.yml]: There is no other data in cohorts other than users, but I can do this for extensibility

**Review by alinjo — APPROVED** [review]

**Review by thomaafl — APPROVED** [review]
schpa

**Pull request merged**